pkg/etcd/etcd.go

```diff
@@ -853,10 +861,152 @@ func (e *ETCD) Snapshot(ctx context.Context, config *config.Control) error {
+	Size      int64        `json:"size,omitempty"`
+	S3        *s3Config    `json:"s3Config,omitempty"`
+}
+
+// listSnapshots provides a list of the currently stored
+// snapshots on disk or in S3 along with their relevant
+// metadata.
+func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]snapshotFile, error) {
+	nodeName := os.Getenv("NODE_NAME")
+
+	var snapshots []snapshotFile
+
+	if e.config.EtcdS3 {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		objects := e.s3.client.ListObjects(ctx, e.config.EtcdS3BucketName, minio.ListObjectsOptions{
+			Prefix: e.s3.snapshotPrefix(),
+		})
+
+		for obj := range objects {
+			if obj.Err != nil {
+				return nil, obj.Err
+			}
+			ca, err := time.Parse(time.RFC3339, obj.LastModified.Format(time.RFC3339))
+			if err != nil {
+				return nil, err
+			}
+			snapshots = append(snapshots, snapshotFile{
+				Name:     obj.Key,
+				NodeName: nodeName,
+				CreatedAt: &metav1.Time{
+					Time: ca,
+				},
+				Size: obj.Size,
+				S3: &s3Config{
+					Endpoint:      e.config.EtcdS3Endpoint,
+					EndpointCA:    e.config.EtcdS3EndpointCA,
+					SkipSSLVerify: e.config.EtcdS3SkipSSLVerify,
+					Bucket:        e.config.EtcdS3BucketName,
+					Region:        e.config.EtcdS3Region,
+					Folder:        e.config.EtcdS3Folder,
+				},
+			})
+		}
+
+		return snapshots, nil
+	}
+
+	files, err := ioutil.ReadDir(snapshotDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		snapshots = append(snapshots, snapshotFile{
+			Name:     f.Name(),
+			Location: "file://" + filepath.Join(snapshotDir, f.Name()),
+			NodeName: nodeName,
+			CreatedAt: &metav1.Time{
+				Time: f.ModTime(),
+			},
+			Size: f.Size(),
+		})
+	}
+
+	return snapshots, nil
+}
+
+// updateSnapshotData populates the given map with the contents of the given slice.
+func updateSnapshotData(data map[string]string, snapshotFiles []snapshotFile) error {
+	for _, v := range snapshotFiles {
+		b, err := json.Marshal(v)
+		if err != nil {
+			return err
+		}
+		data[v.Name] = string(b)
+	}
+
 	return nil
 }
 
-// setSnapshotFunction schedules snapshots at the configured interval
+// storeSnapshotData stores the given snapshot data in the "snapshots" ConfigMap.
+func (e *ETCD) storeSnapshotData(ctx context.Context, snapshotFiles []snapshotFile) error {
+	return retry.OnError(retry.DefaultBackoff, func(err error) bool {
+		return apierrors.IsConflict(err) || apierrors.IsAlreadyExists(err)
+	}, func() error {
+		data := make(map[string]string, len(snapshotFiles))
+		if err := updateSnapshotData(data, snapshotFiles); err != nil {
+			return err
+		}
+
+		// make sure the core.Factory is initialize. There can
+		// be a race between this core code startup.
+		for e.config.Runtime.Core == nil {
+			runtime.Gosched()
+		}
+
+		snapshotConfigMap, err := e.config.Runtime.Core.Core().V1().ConfigMap().Get(metav1.NamespaceSystem, snapshotConfigMapName, metav1.GetOptions{})
```

Comment:
Probably worth assigning `e.config.Runtime.Core.Core().V1().ConfigMap()` to a variable once it's been confirmed that Runtime.Core is non-nil?

Reply:
More a matter of style in that regard over function. I don't have a strong opinion either way.